torch/_refs/nn/functional/__init__.py

```diff
@@ -240,6 +242,35 @@ def softplus(
     return torch.where(scaled_input > threshold, a, rhs)
 
 
+@out_wrapper
+def hardshrink(a: TensorLikeType, lambd: float = 0.5):
+    # Formula for reference,
+    # hardshrink(x) = x if x > lambd
+    #               = x if x < -lambd
+    #               = 0 otherwise
+    zero_mask = refs.logical_and(refs.ge(a, -lambd), refs.le(a, lambd))
+    return refs.where(zero_mask, 0, a)
+
+
+@out_wrapper
+def softshrink(a: TensorLikeType, lambd: float = 0.5):
+    # Formula for reference,
+    # softshrink(x) = x - lambd if x > lambd
+    #               = x + lambd if x < -lambd
+    #               = 0 otherwise
+    ge_mask = refs.ge(a, lambd)
+    a = refs.where(ge_mask, a - lambd, a)
+    le_mask = refs.le(a, -lambd)
+    a = refs.where(le_mask, a + lambd, a)
```

Comment:
Here we may want to update a different object than `a`, because this creates a (false) dependency between the first `where` and the second `where`

Reply:
Also let's bias towards torch calls (because they can be converted to ref calls easily under a context):

```
ge_mask = a > lambd
le_mask = a < lambd
zero_mask = torch.logical_not(torch.logical_or(ge, mask, le_mask)
result = refs.where(ge_mask, a - lambd, a)
result = refs.where(le_mask, a + lambd, a)
result = refs.where(zero_mask, 0, a)
```

?

Reply:
Oh you're right, my implementation was a cheat.

Yep, let's go with your suggestion for now. It'll be interesting to look at how nvFuser fuses this cc @kevinstephano

Reply:
Done! (Had to add an expectedFailure on `test_python_ref` for using `torch.logical_not`